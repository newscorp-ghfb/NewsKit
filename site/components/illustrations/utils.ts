import {SVGAttributes} from 'react';
import {Theme} from '../../../src/theme';

// @ts-ignore
export const toCSSVar = (theme: Theme, props) => {
  const inlineStyle = {} as SVGAttributes<SVGPathElement>;
  const newProps = {...props};

  if (theme.colors[props?.fill!]) {
    inlineStyle.fill = `var(--color-${props.fill})`;
    delete newProps.fill;
  }
  if (theme.colors[props?.stroke!]) {
    inlineStyle.stroke = `var(--color-${props.fill})`;
    delete newProps.stroke;
  }
  return {
    style: inlineStyle,
    ...newProps,
  };
};

const escapeRegExp = (string: string): string =>
  // $& means the whole matched string
  string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

export const pathToID = (path: string) => {
  if (path) {
    return path
      .replace(new RegExp(escapeRegExp('../'), 'g'), '')
      .replace(new RegExp(escapeRegExp('./'), 'g'), '')
      .replace(new RegExp(escapeRegExp('/'), 'g'), '-');
  }
  return '';
};

export const isSafari =
  typeof window !== 'undefined' &&
  /^((?!chrome|android).)*safari/i.test(window.navigator.userAgent);
