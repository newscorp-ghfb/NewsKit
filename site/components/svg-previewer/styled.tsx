import {Button, styled, GridLayout, getColorCssFromTheme} from 'newskit';

export const StyledSingleSVGDownloadButton = styled(Button)`
  margin-right: 20px;
`;

export const StyledSvgGroupContainer = styled(GridLayout)``;

export const StyledSvgPreviewerContainer = styled(GridLayout)`
  /*
  display: flex;
  flex-direction: column;
  align-items: center;
  */
`;

export const StyledButtonsContainer = styled(GridLayout)`
  position: fixed;
  top: var(--headerSize);
  margin-top: -12px;
  ${getColorCssFromTheme('background', 'interface020')}
  ${getColorCssFromTheme('borderBottomColor', 'interface040')}
  border-bottom-width: 1px;
  border-bottom-style: solid;
  z-index: 1000;
`;

export const StyledSingleSvgWrapper = styled.div`
  padding: 0 10vw;
  svg {
    max-width: 80vw;
    max-height: 80vh;
  }
`;
