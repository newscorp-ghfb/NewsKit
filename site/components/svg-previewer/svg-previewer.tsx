import React, {useEffect, useState} from 'react';
import {
  getColorCssFromTheme,
  Scroll,
  P,
  ScrollSnapAlignment,
  GridLayout,
} from 'newskit';
import dompurify from 'dompurify';
import {themeList, ThemeNames} from './colors-theme-list';
import {StyledButtonsContainer, StyledSingleSvgWrapper} from './styled';
import {DownloadControls} from './controls/download-controls';
import {ThemeControls} from './controls/theme-controls';

export const SvgPreviewer: React.FC = () => {
  const sanitizer = dompurify.sanitize;

  // The name of the theme being currently used, the initial value will decide the default theme.
  const [currentThemeName, setCurrentThemeName] = useState<ThemeNames>(
    'Docs Theme',
  );

  // SVG coming from Figma - used as a base for some mutations.
  const [baseSvgCodeGroup, setBaseSvgCodeGroup] = useState<
    [{figmaSvg: string; name: string}]
  >();

  // SVGs to be displayed in the page
  const [svgCodeGroup, setSvgCodeGroup] = useState<[string]>();

  // Hexes obj sent by figma, used for switch theme logic.
  const [hexesObj, setHexesObj] = useState<Object>({});

  // const setIds = (figmaSvg: string) => {
  //   const regexList = [/mask\d{1,}/g, /filter\d{1,}/g, /clip\d{1,}/g];
  //   let figmaSvgCopy = figmaSvg;

  //   regexList.forEach(regex => {
  //     const matchList = figmaSvgCopy.match(regex);

  //     matchList?.forEach(match => {
  //       figmaSvgCopy = figmaSvgCopy.replace(
  //         new RegExp(match, 'g'),
  //         `${getSSRId()}`,
  //       );
  //     });
  //   });
  //   return figmaSvgCopy;
  // };

  const testSvgCodeIsString = (data: Array<{code: string; name: string}>) => {
    data.forEach(svg => {
      if (typeof svg.code !== 'string') throw new Error();
    });
  };

  // // This function simplify the IDs set by figma for clips, filters and masks.
  // // It makes simpler the SVG manipulation especially when building the react component.
  // // ** To be removed if Figma brings back short IDs (E.G: clip0 than clip0_12:1464). **
  // const svgIdsShortener = (figmaSvg: string) => {
  //   let figmaSvgCopy = figmaSvg;
  //   const regexList = [
  //     /mask\d+_\d+_\d+/g,
  //     /filter\d+_d_\d+_\d+/g,
  //     /clip\d+_\d+_\d+/g,
  //   ];

  //   regexList.forEach(regex => {
  //     const matchList = figmaSvg.match(regex);

  //     matchList?.forEach(match => {
  //       const firstUnderscoreIndex = match.indexOf('_');

  //       figmaSvgCopy = figmaSvgCopy.replace(
  //         new RegExp(match, 'g'),
  //         match.substring(0, firstUnderscoreIndex),
  //       );
  //     });
  //   });
  //   return figmaSvgCopy;
  // };

  const getThemeFromList = (
    themeName: ThemeNames,
    version: 'light' | 'dark',
  ) => {
    const newTheme = themeList.filter(theme => theme.name === themeName);

    if (version === 'light') {
      return newTheme[0].defaultLight;
    }
    return newTheme[0].dark;
  };

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const onmessage = (event: MessageEvent) => {
    if (event.data.pluginMessage?.action === 'FilesToUI') {
      testSvgCodeIsString(event.data.pluginMessage.data.svgdata);

      // Left console logs as can be useful for debugging SVG issues without
      // using the local version.
      // eslint-disable-next-line no-console
      console.log(event.data.pluginMessage.data.hexes, 'Hexes');
      // eslint-disable-next-line no-console
      console.log(event.data.pluginMessage.data, 'SVG data');

      setHexesObj(event.data.pluginMessage.data.hexes);

      const baseFigmaSvg: Array<{figmaSvg: string; name: string}> = [];
      // @ts-ignore
      const content = event.data.pluginMessage.data.svgdata.map(svg => {
        let figmaSvg: string = svg.code;
        // figmaSvg = svgIdsShortener(figmaSvg);
        baseFigmaSvg.push({figmaSvg, name: svg.name});

        // Setting ids for "mask", "filter", and "clip" attributes
        // figmaSvg = setIds(figmaSvg);

        // ** REPLACING RANDOM FIGMA COLORS
        Object.entries(event.data.pluginMessage.data.hexes).forEach(hex => {
          const colorToken = hex[1];
          const figmaRandomColour = hex[0];
          const defaultTheme = getThemeFromList(currentThemeName, 'light')!;

          const {lightThemeColorValue} = getColorCssFromTheme(
            'lightThemeColorValue',
            colorToken as string,
          )({theme: defaultTheme}) as {
            lightThemeColorValue: string;
          };

          // Replace hardcoded figma #colour with light theme token value
          figmaSvg = figmaSvg?.replace(
            new RegExp(figmaRandomColour, 'gim'),
            lightThemeColorValue,
          );
        });
        //  **

        return figmaSvg;
      });

      // @ts-ignore
      setBaseSvgCodeGroup(baseFigmaSvg);
      setSvgCodeGroup(content);
    }
  };

  useEffect(() => {
    window.addEventListener('message', onmessage);

    return () => {
      window.removeEventListener('message', onmessage);
    };
  }, [onmessage]);

  return (
    <GridLayout>
      <StyledButtonsContainer
        rows={{xs: 'sizing080'}}
        alignItems="center"
        justifyContent="sapce-between"
        columns="1fr 1fr"
        overrides={{
          paddingInline: '10%',
          paddingBlock: '8px',
          minWidth: '100%',
        }}
      >
        <GridLayout
          columns="1fr 1fr"
          alignItems="center"
          rows={{xs: 'sizing080'}}
        >
          <ThemeControls
            setSvgCodeGroup={setSvgCodeGroup}
            hexesObj={hexesObj}
            baseSvgCodeGroup={baseSvgCodeGroup}
            svgCodeGroup={svgCodeGroup}
            getThemeFromList={getThemeFromList}
            currentThemeName={currentThemeName}
            setCurrentThemeName={setCurrentThemeName}
          />
        </GridLayout>
        <GridLayout
          columns="1fr 1fr 1fr"
          alignItems="baseline"
          rows={{xs: 'sizing080'}}
        >
          <DownloadControls
            hexesObj={hexesObj}
            baseSvgCodeGroup={baseSvgCodeGroup}
            svgCodeGroup={svgCodeGroup}
          />
        </GridLayout>
      </StyledButtonsContainer>

      {svgCodeGroup && baseSvgCodeGroup && (
        <Scroll
          id="scroll"
          snapAlign="center"
          controls="static"
          stepDistance={window.innerWidth}
        >
          <GridLayout
            autoRows="1fr"
            autoFlow="column dense"
            overrides={{
              marginBlock: '10vh',
              maxHeight: '80vh',
            }}
          >
            {svgCodeGroup.map((svgCode, index) => (
              <ScrollSnapAlignment key={baseSvgCodeGroup[index].name}>
                <StyledSingleSvgWrapper>
                  <P>{baseSvgCodeGroup[index].name}</P>
                  {/* eslint-disable-next-line react/no-danger */}
                  <div dangerouslySetInnerHTML={{__html: sanitizer(svgCode)}} />
                </StyledSingleSvgWrapper>
              </ScrollSnapAlignment>
            ))}
          </GridLayout>
        </Scroll>
      )}
    </GridLayout>
  );
};
