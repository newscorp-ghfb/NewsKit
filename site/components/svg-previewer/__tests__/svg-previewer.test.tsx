import {act, fireEvent} from '@testing-library/react';
import '@testing-library/jest-dom';
import FileSaver from 'file-saver';
import {renderWithTheme} from '../../../utils/test-utils';
import {SvgPreviewer} from '../svg-previewer';
import {fileSvgMock} from './expected-file-mock';
import {Hexes, SVGString} from './figma-svg-mock-data';

jest.mock('file-saver', () => ({saveAs: jest.fn()}));
// @ts-ignore
global.Blob = function blob(content: any, options: any) {
  return {content, options};
};

describe('SvgPreviewer', () => {
  const delay = (ms: any) => new Promise(res => setTimeout(res, ms));
  const mockResizeObserver = jest.fn(() => ({
    observe: jest.fn(),
    disconnect: jest.fn(),
  }));

  beforeAll(() => {
    // @ts-ignore
    global.ResizeObserver = mockResizeObserver;
  });

  afterAll(() => {
    // @ts-ignore
    global.ResizeObserver = null;
  });

  it('should render correctly when not receiving any data', async () => {
    const {asFragment, findByTestId} = renderWithTheme(SvgPreviewer);
    const data = {};

    await act(() => {
      window.postMessage(data, '*');
    });

    const selectButton = await findByTestId('select-theme-element');

    expect(selectButton).toBeDisabled();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly when receiving one svg data', async () => {
    const {asFragment, findByTestId} = renderWithTheme(SvgPreviewer);
    const data = {
      pluginMessage: {
        action: 'FilesToUI',
        data: {
          hexes: Hexes,
          svgdata: [{code: SVGString, name: 'SvgName'}],
        },
      },
    };

    await act(async () => {
      window.postMessage({...data}, '*');
      await delay(50);
    });

    const selectButton = await findByTestId('select-theme-element');
    expect(selectButton).not.toBeDisabled();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should compose the SVG download file as expected', async () => {
    const {findByTestId} = renderWithTheme(SvgPreviewer);
    const data = {
      pluginMessage: {
        action: 'FilesToUI',
        data: {
          hexes: Hexes,
          svgdata: [{code: SVGString, name: 'Card No Crop'}],
        },
      },
    };

    await act(async () => {
      window.postMessage({...data}, '*');
      await delay(50);
    });

    const svgButton = await findByTestId('single-svg-button');
    fireEvent.click(svgButton);

    expect(FileSaver.saveAs).toHaveBeenCalledWith(
      {content: [fileSvgMock], options: {type: 'text/plain;charset=utf-8;'}},
      'card-no-crop.svg',
    );
  });

  it('should change theme colors correctly when selecting a different theme from select element', async () => {
    const {findByTestId, getAllByRole, asFragment} = renderWithTheme(
      SvgPreviewer,
    );
    const data = {
      pluginMessage: {
        action: 'FilesToUI',
        data: {
          hexes: Hexes,
          svgdata: [{code: SVGString, name: 'SvgName'}],
        },
      },
    };

    await act(async () => {
      window.postMessage({...data}, '*');
      await delay(50);
    });
    // open select
    const selectButton = await findByTestId('select-theme-element');
    fireEvent.click(selectButton);
    // select 3rd option
    fireEvent.click(getAllByRole('option')[2]);

    expect(asFragment()).toMatchSnapshot();
  });
});
