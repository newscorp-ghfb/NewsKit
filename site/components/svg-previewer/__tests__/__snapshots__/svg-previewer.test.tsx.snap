// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SvgPreviewer should render correctly when not receiving any data 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0;
  padding: 0;
  display: grid;
}

.emotion-1 {
  margin: 0;
  padding: 0;
  display: grid;
  grid-template-columns: 1fr 1fr;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 100%;
  padding-inline: 10%;
  padding-block: 8px;
  position: fixed;
  top: var(--headerSize);
  margin-top: -12px;
  background: #F0F1F3;
  border-bottom-color: #DEE2E5;
  border-bottom-width: 1px;
  border-bottom-style: solid;
  z-index: 1000;
}

@media screen {
  .emotion-1 {
    grid-template-rows: 48px;
  }
}

.emotion-2 {
  margin: 0;
  padding: 0;
  display: grid;
  grid-template-columns: 1fr 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media screen {
  .emotion-2 {
    grid-template-rows: 48px;
  }
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  max-width: 196px;
  min-height: 48px;
  margin-bottom: 8px;
  background-color: transparent;
  border-style: solid;
  border-color: #DEE2E5;
  border-width: 1px;
  border-radius: 8px;
  color: #C0C7CC;
  text-overflow: ellipsis;
  margin-block: 0px;
}

.emotion-3 svg {
  fill: #C0C7CC;
}

.emotion-4 {
  display: block;
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  margin: -1px;
}

.emotion-5 {
  all: unset;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  width: 100%;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  overflow: hidden;
  cursor: not-allowed;
  padding: 8px;
}

.emotion-6 {
  width: 100%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  color: inherit;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0;
}

.emotion-7 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 8px;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-column-gap: 8px;
  column-gap: 8px;
  cursor: not-allowed;
}

.emotion-9 {
  all: unset;
}

.emotion-10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-11 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
  fill: #2E3F54;
  vertical-align: unset;
  display: inline-block;
}

@media screen and (prefers-reduced-motion: no-preference) {
  .emotion-11 {
    transition-property: fill;
    transition-duration: 200ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1);
  }
}

@media screen and (prefers-reduced-motion: reduce) {
  .emotion-11 {
    transition-property: fill;
    transition-duration: 0ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1);
  }
}

.emotion-11.emotion-11 {
  width: 24px;
  height: 24px;
}

.emotion-13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.emotion-14 {
  position: relative;
  display: inline-block;
  -webkit-align-self: start;
  -ms-flex-item-align: start;
  align-self: start;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  block-size: 32px;
  inline-size: 60px;
  margin-right: 12px;
}

.emotion-14 * {
  box-sizing: border-box;
}

.emotion-15 {
  position: absolute;
  top: 50%;
  left: calc(100% - (24px / 2) - 4px);
  background-color: #3358CC;
  border-radius: 50%;
  opacity: 0;
  width: 1px;
  height: 1px;
  -webkit-transform: translate3d(calc(48px / -2), calc(48px / -2), 0);
  -moz-transform: translate3d(calc(48px / -2), calc(48px / -2), 0);
  -ms-transform: translate3d(calc(48px / -2), calc(48px / -2), 0);
  transform: translate3d(calc(48px / -2), calc(48px / -2), 0);
}

.emotion-15:disabled {
  opacity: 0;
}

.emotion-15:hover:active:not(:disabled) {
  opacity: 0.4;
}

.emotion-15:valid:hover:not(:disabled) {
  background-color: #00601A;
  opacity: 0.2;
}

.emotion-15:valid:hover:active:not(:disabled) {
  background-color: #00601A;
  opacity: 0.4;
}

.emotion-15:invalid:hover:not(:disabled) {
  background-color: #A90000;
  opacity: 0.2;
}

.emotion-15:invalid:hover:active:not(:disabled) {
  background-color: #A90000;
  opacity: 0.4;
}

@media screen and (prefers-reduced-motion: no-preference) {
  .emotion-15 {
    transition-property: inset,opacity;
    transition-duration: 200ms,200ms;
    transition-timing-function: cubic-bezier(.5, 0, 1, 1),cubic-bezier(0, 0, .5, 1);
  }
}

@media screen and (prefers-reduced-motion: reduce) {
  .emotion-15 {
    transition-property: inset,opacity;
    transition-duration: 0ms;
    transition-timing-function: cubic-bezier(.5, 0, 1, 1),cubic-bezier(0, 0, .5, 1);
  }
}

.emotion-16 {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #3358CC;
  border-radius: 20rem;
  padding-inline: 4px;
}

.emotion-16:focus-visible:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  outline-offset: 2px;
  background-color: #3358CC;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-16:focus-visible:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-17 {
  position: relative;
  width: 100%;
  height: 100%;
  display: grid;
  grid-template-columns: 1fr 1fr;
  place-items: center;
}

.emotion-18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
}

.emotion-18 svg {
  fill: #FFFFFF;
}

.emotion-18:disabled svg {
  fill: #C0C7CC;
}

.emotion-18 svg {
  height: 100%;
  width: 100%;
  max-width: 24px;
  max-height: 32px;
}

.emotion-19 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
  fill: #2E3F54;
  vertical-align: unset;
  display: inline-block;
}

@media screen and (prefers-reduced-motion: no-preference) {
  .emotion-19 {
    transition-property: fill;
    transition-duration: 200ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1);
  }
}

@media screen and (prefers-reduced-motion: reduce) {
  .emotion-19 {
    transition-property: fill;
    transition-duration: 0ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1);
  }
}

.emotion-21 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
}

.emotion-21 svg {
  fill: #FFFFFF;
}

.emotion-21:disabled svg {
  fill: #C0C7CC;
}

.emotion-21 svg {
  height: 100%;
  width: 100%;
  max-width: 24px;
  max-height: 32px;
}

.emotion-24 {
  background-color: #FFFFFF;
  border-radius: 50%;
  width: 24px;
  height: 24px;
  left: calc(100% - 24px);
  padding-inline: 4px;
  padding-block: 4px;
  position: absolute;
  aspect-ratio: 1/1;
  display: grid;
  place-items: center;
}

.emotion-24 svg {
  fill: #09111C;
}

@media screen and (prefers-reduced-motion: no-preference) {
  .emotion-24 {
    transition-property: inset;
    transition-duration: 200ms;
    transition-timing-function: cubic-bezier(.5, 0, 1, 1);
  }
}

@media screen and (prefers-reduced-motion: reduce) {
  .emotion-24 {
    transition-property: inset;
    transition-duration: 0ms;
    transition-timing-function: cubic-bezier(.5, 0, 1, 1);
  }
}

.emotion-25 {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  margin: 0;
  opacity: 0;
  cursor: inherit;
}

.emotion-26 {
  color: #2E3F54;
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
}

.emotion-26:disabled {
  color: #C0C7CC;
}

.emotion-26::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-26::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-27 {
  margin: 0;
  padding: 0;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

@media screen {
  .emotion-27 {
    grid-template-rows: 48px;
  }
}

.emotion-28 {
  margin: 0;
  padding: 0;
  display: inline-grid;
  -webkit-column-gap: 8px;
  column-gap: 8px;
  grid-template-columns: repeat(1, auto);
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  justify-items: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 12px 16px;
  min-width: 80px;
  min-height: 48px;
  box-sizing: border-box;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 48px;
  min-width: 80px;
  cursor: default;
  overflow: hidden;
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: transparent;
  border-radius: 8px;
  color: #3358CC;
  cursor: pointer;
  margin-right: 20px;
}

.emotion-28 svg {
  width: 24px;
  height: 24px;
}

.emotion-28 svg {
  fill: #3358CC;
}

.emotion-28:hover:not(:disabled) {
  background-color: #ECF1FF;
}

.emotion-28:active:not(:disabled) {
  background-color: #D5E0FC;
}

.emotion-28:disabled {
  color: #C0C7CC;
}

.emotion-28:disabled svg {
  fill: #C0C7CC;
}

.emotion-28:focus-visible:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  outline-offset: 2px;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-28:focus-visible:not(:disabled) {
      outline-style: auto;
    }
  }
}

@media screen and (prefers-reduced-motion: no-preference) {
  .emotion-28 {
    transition-property: background-color;
    transition-duration: 200ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1);
  }
}

@media screen and (prefers-reduced-motion: reduce) {
  .emotion-28 {
    transition-property: background-color;
    transition-duration: 0ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1);
  }
}

.emotion-29 {
  margin: 0;
  font-family: "Poppins",sans-serif;
  font-size: 14px;
  line-height: 21px;
  font-weight: 600;
  letter-spacing: 0;
  padding: 0.5px 0px;
  display: inline-block;
}

.emotion-29::before {
  content: '';
  margin-bottom: -0.403em;
  display: block;
}

.emotion-29::after {
  content: '';
  margin-top: -0.4em;
  display: block;
}

.emotion-30 {
  margin: 0;
  padding: 0;
  display: inline-grid;
  -webkit-column-gap: 8px;
  column-gap: 8px;
  grid-template-columns: repeat(1, auto);
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  justify-items: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 12px 16px;
  min-width: 80px;
  min-height: 48px;
  box-sizing: border-box;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 48px;
  min-width: 80px;
  cursor: default;
  overflow: hidden;
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: #3358CC;
  border-radius: 8px;
  color: #FFFFFF;
  cursor: pointer;
}

.emotion-30 svg {
  width: 24px;
  height: 24px;
}

.emotion-30 svg {
  fill: #FFFFFF;
}

.emotion-30:hover:not(:disabled) {
  background-color: #254CAC;
}

.emotion-30:active:not(:disabled) {
  background-color: #14328D;
}

.emotion-30:disabled {
  background-color: #DEE2E5;
  color: #C0C7CC;
}

.emotion-30:disabled svg {
  fill: #C0C7CC;
}

.emotion-30:focus-visible:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  outline-offset: 2px;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-30:focus-visible:not(:disabled) {
      outline-style: auto;
    }
  }
}

@media screen and (prefers-reduced-motion: no-preference) {
  .emotion-30 {
    transition-property: background-color;
    transition-duration: 200ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1);
  }
}

@media screen and (prefers-reduced-motion: reduce) {
  .emotion-30 {
    transition-property: background-color;
    transition-duration: 0ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1);
  }
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-3"
        >
          <span
            class="emotion-4"
          />
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            class="emotion-5"
            data-testid="select-theme-element"
            disabled=""
            id="downshift-0-toggle-button"
            type="button"
            value="Docs Theme"
          >
            <div
              class="emotion-6"
              disabled=""
            >
              Select Option
            </div>
          </button>
          <div
            class="emotion-7"
          >
            <div
              class="emotion-8"
            >
              <button
                aria-hidden="true"
                class="emotion-9"
                data-testid="select-chevron-button"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <div
                  class="emotion-10"
                >
                  <svg
                    aria-hidden="true"
                    class="emotion-11 emotion-12"
                    fill="currentColor"
                    focusable="false"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0V0z"
                      fill="none"
                    />
                    <path
                      d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                    />
                  </svg>
                </div>
              </button>
            </div>
          </div>
        </div>
        <label
          class="emotion-13"
        >
          <div
            class="emotion-14"
            role="presentation"
          >
            <div
              class="emotion-15"
              data-testid="checkbox-feedback"
            />
            <div
              class="emotion-16"
            >
              <div
                class="emotion-17"
              >
                <div
                  class="emotion-18"
                >
                  <svg
                    aria-hidden="true"
                    class="emotion-19 emotion-12"
                    fill="currentColor"
                    focusable="false"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <rect
                      fill="none"
                      height="24"
                      width="24"
                    />
                    <path
                      d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zM2 13h2c.55 0 1-.45 1-1s-.45-1-1-1H2c-.55 0-1 .45-1 1s.45 1 1 1zm18 0h2c.55 0 1-.45 1-1s-.45-1-1-1h-2c-.55 0-1 .45-1 1s.45 1 1 1zM11 2v2c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1s-1 .45-1 1zm0 18v2c0 .55.45 1 1 1s1-.45 1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1zM5.99 4.58a.996.996 0 00-1.41 0 .996.996 0 000 1.41l1.06 1.06c.39.39 1.03.39 1.41 0s.39-1.03 0-1.41L5.99 4.58zm12.37 12.37a.996.996 0 00-1.41 0 .996.996 0 000 1.41l1.06 1.06c.39.39 1.03.39 1.41 0a.996.996 0 000-1.41l-1.06-1.06zm1.06-10.96a.996.996 0 000-1.41.996.996 0 00-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06zM7.05 18.36a.996.996 0 000-1.41.996.996 0 00-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06z"
                    />
                  </svg>
                </div>
                <div
                  class="emotion-21"
                >
                  <svg
                    aria-hidden="true"
                    class="emotion-19 emotion-12"
                    fill="currentColor"
                    focusable="false"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <rect
                      fill="none"
                      height="24"
                      width="24"
                    />
                    <path
                      d="M12 3a9 9 0 109 9c0-.46-.04-.92-.1-1.36a5.389 5.389 0 01-4.4 2.26 5.403 5.403 0 01-3.14-9.8c-.44-.06-.9-.1-1.36-.1z"
                    />
                  </svg>
                </div>
                <div
                  class="emotion-24"
                />
              </div>
              <input
                checked=""
                class="emotion-25"
                data-testid="checkbox-input"
                role="switch"
                tabindex="-1"
                type="checkbox"
              />
            </div>
          </div>
          <span
            class="emotion-26"
          >
            Toggle Theme
          </span>
        </label>
      </div>
      <div
        class="emotion-27"
      >
        <button
          class="emotion-28"
          data-testid="single-svg-button"
          type="button"
        >
          <span
            class="emotion-29"
          >
            Download .svg
          </span>
        </button>
        <button
          class="emotion-30"
          data-testid="button"
          type="button"
        >
          <span
            class="emotion-29"
          >
            Download all .zip
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
